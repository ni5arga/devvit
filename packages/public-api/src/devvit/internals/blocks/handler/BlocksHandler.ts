import {
  EffectType,
  type Effect,
  type Metadata,
  type UIEvent,
  type UIRequest,
  type UIResponse,
} from '@devvit/protos';
import { ContextBuilder } from './ContextBuilder.js';
import { BlocksTransformer } from '../BlocksTransformer.js';
import type { BlockElement } from '../../../Devvit.js';
import type { ReifiedBlockElement, ReifiedBlockElementOrLiteral } from '../BlocksReconciler.js';
import type { Hook, HookSegment, HookParams, Props, BlocksState } from './types.js';
import { RenderInterruptError } from './types.js';
import { RenderContext } from './RenderContext.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import type { EffectEmitter } from '../EffectEmitter.js';

/**
 * This can be a global/singleton because render is synchronous.
 *
 * If you want to use this from somewhere else, please consider using one of the
 * functions like isRendering or registerHook, and then try to add additional
 * functions here if needed.  Don't use this directly.
 */
export let _activeRenderContext: RenderContext | null = null;

export function useEffectEmitter(): EffectEmitter {
  if (!_activeRenderContext) {
    throw new Error('Hooks can only be declared at the top of a component.');
  }
  return _activeRenderContext;
}

export function isRendering(): boolean {
  return _activeRenderContext !== null;
}

function _structuredClone<T extends JSONValue>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * This is the recommended low-level interface for creating hooks like useState or useAsync.
 *
 * Practically, this initializes your hook if it doesn't already exist, and makes sure
 * that its state gets all sync'd up.
 *
 * @param HookSegment -- A name for this hook.  This is used to dedupe hooks.
 * @param initializer
 *    factory for building this hook
 * @returns
 */
export function registerHook<H extends Hook>(
  options: HookSegment,
  initializer: (p: HookParams) => H
): H {
  if (!_activeRenderContext) {
    throw new Error(
      "Hooks can only be declared at the top of a component.  You cannot declare hooks outside\
    of components or inside of event handlers.  It's almost always a mistake to declare hooks inside of loops or\
    conditionals."
    );
  }
  const hookId = _activeRenderContext.nextHookId(options);
  const context = _activeRenderContext;
  const params: HookParams = {
    hookId,
    changed: () => {
      context._changed[hookId] = true;
      context._state[hookId] = context?._hooks[hookId]?.state;
    },
    context: _activeRenderContext,
  };
  const fromNull = _activeRenderContext._state[hookId] === undefined;
  _activeRenderContext._hooks[hookId] = _activeRenderContext._hooks[hookId] ?? initializer(params);
  const hook: H = _activeRenderContext._hooks[hookId] as H;

  if (_activeRenderContext._state[hookId] !== undefined) {
    hook.state = _activeRenderContext._state[hookId];
  }
  if (hook.onLoad) {
    hook.onLoad(_activeRenderContext);
  }
  if (fromNull && hook.state !== undefined && hook.state !== null) {
    params.changed();
  }
  return hook;
}

export let _latestBlocksHandler: BlocksHandler | null = null;

/**
 * Replacing BlocksReconciler, the model is now less of a "reconciliation", and more
 * of a handling a request/response lifecycle.
 *
 */
export class BlocksHandler {
  #root: JSX.ComponentFunction;
  #contextBuilder: ContextBuilder = new ContextBuilder();
  #blocksTransformer: BlocksTransformer = new BlocksTransformer();
  _latestRenderContext: RenderContext | null = null;

  constructor(root: JSX.ComponentFunction) {
    this.#root = root;
    _latestBlocksHandler = this;
  }

  async handle(request: UIRequest, metadata?: Metadata): Promise<UIResponse> {
    const context = new RenderContext(request);
    const devvitContext = this.#contextBuilder.buildContext(context, metadata!);
    context.devvitContext = devvitContext;

    let blocks;

    /**
     * Events on the main queue must be handled in order, so that state is updated in the correct order.  Events
     * on other queues can be handled in parallel, because they only emit effects.
     *
     * There is an optimization here to process SendEventEffects locally, instead of letting them bubble up to the
     * platform.  This prevents a round trip to the platform for every event.
     *
     * This also means we need to respect execution queues here, and not just in the platform.
     */
    const eventsToProcess = request.events;
    const noEvents = !request.events?.length;
    const isMainQueue = noEvents || eventsToProcess.some((e) => !e.async);

    const isBlockingSSR = eventsToProcess.some((e) => e.blocking);

    let progress:
      | {
          _state: BlocksState;
          _effects: { [key: string]: Effect };
        }
      | undefined;
    let remaining: UIEvent[] = [...eventsToProcess];

    while (eventsToProcess.length > 0) {
      /**
       * A concurrently executable batch is a set of events that can be executed in parallel.  This either one main queue event,
       * or any number of other queue events.
       */
      const batch = [];
      if (!eventsToProcess[0].async) {
        batch.push(eventsToProcess.shift()!);
      } else {
        while (eventsToProcess[0]?.async) {
          batch.push(eventsToProcess.shift()!);
        }
      }
      if (!batch.length) throw Error('batch must have at least one event');
      try {
        if (batch[0].async) {
          const stateCopy = _structuredClone(context._state);
          await this.#handleAsyncQueues(context, ...batch);
          // enforce that state updates are only allowed on the main queue.
          context._state = stateCopy;
        } else {
          await this.#handleMainQueue(context, ...batch);
        }
      } catch (e) {
        /**
         * If we have a progress, we can recover from an error by rolling back to the last progress, and then letting the
         * remaining events be reprocessed.
         */
        if (progress) {
          context._state = progress._state;
          context._effects = progress._effects;
          remaining.forEach((e, i) => {
            const effect: Effect = {
              type: EffectType.EFFECT_SEND_EVENT,
              sendEvent: {
                event: e,
                jumpsQueue: true,
              },
            };
            context.emitEffect(`remaining-${i}`, effect);
          });
          break;
        } else {
          throw e;
        }
      }

      /**
       * If we have any SendEventEffects, we can push them back on the queue to process them locally
       */
      for (const [key, effect] of Object.entries(context._effects)) {
        if (effect.sendEvent?.event) {
          if (!isMainQueue && !effect.sendEvent?.event?.async) {
            // We're async, this is a main queue event.  We need to send it back to the platform to let
            // the platform synchronize it.
            break;
          }

          if (isMainQueue && effect.sendEvent?.event?.async && !isBlockingSSR) {
            // We're main queue, and this is an async event.  We're not in SSR mode, so let's prioritize
            // returning control quickly to the platform so we don't block event loops.
            break;
          }

          //Ok, we can handle this event locally.
          const event = effect.sendEvent.event;
          eventsToProcess.push(event);
          delete context._effects[key];
        }
      }

      /**
       * If we're going back through this again, we need to capture the progress, and the remaining events.
       */
      if (eventsToProcess.length > 0) {
        progress = {
          _state: _structuredClone(context._state),
          _effects: { ...context._effects },
        };
        remaining = [...eventsToProcess];
      }
    } // End of while loop

    if (isMainQueue) {
      // Rendering only happens on the main queue.
      const tags = this.#renderRoot(this.#root, context._rootProps ?? {}, context);
      if (tags) {
        blocks = await this.#blocksTransformer.createBlocksElementOrThrow(tags);
        blocks = await this.#blocksTransformer.ensureRootBlock(blocks);
      }
    }

    return {
      state: context._state,
      effects: context.effects,
      blocks,
    };
  }

  #loadHooks(context: RenderContext, ..._events: UIEvent[]): void {
    // TBD: partial rendering
    this.#renderRoot(this.#root, context.request.props ?? {}, context);
  }

  /**
   * These can all run in parallel, because they only emit effects
   */
  async #handleAsyncQueues(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    this.#loadHooks(context, ...batch);

    await Promise.all(
      batch.map(async (event) => {
        if (!event.async) {
          throw new Error(
            "You can't mix main and other queues in one batch.  This is likely a platform bug.  Please file an issue in the Discord for someone to help! https://discord.com/channels/1050224141732687912/1115441897079574620"
          );
        }
        await this.#attemptHook(context, event);
      })
    );
  }

  async #attemptHook(context: RenderContext, event: UIEvent): Promise<void> {
    const hook = context._hooks[event.hook!];
    if (hook && hook.onUIEvent) {
      try {
        await hook.onUIEvent(event, context);
      } catch (e) {
        console.error('Error in event handler', e);
        throw e;
      }
    } else {
      await context.handleUndeliveredEvent(event);
    }
  }

  async #handleMainQueue(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    // We need to handle events in order, so that the state is updated in the correct order.
    for (const event of batch) {
      this.#loadHooks(context, event);
      context._state.__generation = Number(context._state.__generation ?? 0) + 1;
      await this.#attemptHook(context, event);
    }
  }

  #renderRoot(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElement | undefined {
    context._generated = {};
    _activeRenderContext = context;
    this._latestRenderContext = context;
    try {
      const roots = this.#render(component, props, context);
      if (roots.length !== 1) {
        throw new Error('only one root');
      }
      const root = roots[0];
      if (typeof root === 'string') {
        throw new Error(
          'There must be a root tag.  Try wrapping your app in a <text></text>, <vstack> or other tag.'
        );
      }
      return root;
    } catch (e) {
      if (e instanceof RenderInterruptError) {
        return undefined;
      } else {
        throw e;
      }
    } finally {
      _activeRenderContext = null;
    }
  }

  #render(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElementOrLiteral[] {
    context.push({ namespace: component.name, ...props });
    try {
      const element = component(props, context.devvitContext);
      return this.#renderElement(element, context);
    } finally {
      context.pop();
    }
  }

  #renderList(list: JSX.Element[], context: RenderContext): ReifiedBlockElementOrLiteral[] {
    list = list.flat(Infinity);
    return list.flatMap((e, i) => {
      if (e && typeof e === 'object' && 'props' in e) {
        if (!e.props?.key) {
          e.props = e.props ?? {};
          e.props.key = i;
        }
      }
      return this.#renderElement(e, context);
    });
  }

  #renderElement(element: JSX.Element, context: RenderContext): ReifiedBlockElementOrLiteral[] {
    if (Array.isArray(element)) {
      return this.#renderList(element, context);
    } else if (isBlockElement(element)) {
      if (element.type === undefined) {
        return this.#renderList(element.children, context);
      } else if (typeof element.type === 'function') {
        const propsWithChildren = { ...element.props, children: element.children };
        return this.#render(element.type, propsWithChildren, context);
      } else {
        context.push({ namespace: element.type, ...element.props });
        const reifiedChildren = this.#renderList(element.children, context);
        const reifiedProps = this.#reifyProps(element.props ?? {});
        context.pop();
        return [{ type: element.type, children: reifiedChildren, props: reifiedProps }];
      }
    } else {
      return [(element ?? '').toString()];
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  #reifyProps(props: { [key: string]: any }): { [key: string]: string } {
    const reifiedProps: { [key: string]: string } = {};
    for (const key in props) {
      if (typeof props[key] === 'function') {
        const hook = registerHook(
          {
            namespace: key,
            key: false,
          },
          ({ hookId }) => ({ hookId, state: null, onUIEvent: props[key] })
        );
        reifiedProps[key] = hook.hookId;
        if ('captureHookRef' in props[key]) {
          props[key].captureHookRef();
        }
      } else {
        const value = props[key];
        if (value !== undefined && value !== null) {
          reifiedProps[key] = value.toString();
        }
      }
    }
    return reifiedProps;
  }
}

function isBlockElement(e: JSX.Element): e is BlockElement {
  return typeof e === 'object' && e != null && 'type' in e;
}
