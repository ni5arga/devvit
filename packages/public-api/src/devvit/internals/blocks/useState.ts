import type { SetStateAction, UseStateHook, UseStateResult } from '../../../types/hooks.js';
import type { BlocksReconciler } from './BlocksReconciler.js';

export function makeUseStateHook(reconciler: BlocksReconciler): UseStateHook {
  function useState<S>(initialState: S): UseStateResult<S> {
    const hookIndex = reconciler.currentHookIndex;
    const currentState = reconciler.getCurrentComponentState<S>();
    const previousState = reconciler.getPreviousComponentState<S>();

    if (hookIndex in currentState) {
      reconciler.currentHookIndex++;
      return [currentState[hookIndex], stateSetter];
    }

    if (reconciler.isInitialRender || !(hookIndex in previousState)) {
      const value = initialState instanceof Function ? initialState() : initialState;

      if (value instanceof Promise) {
        const asyncResolver = async (): Promise<void> => {
          currentState[hookIndex] = await value;
          reconciler.currentHookIndex = 0;
        };
        throw asyncResolver();
      }

      currentState[hookIndex] = value;
    } else {
      currentState[hookIndex] = previousState[hookIndex];
    }

    function stateSetter(valueOrFunction: SetStateAction<S>): void {
      if (reconciler.isRendering) {
        throw new Error('Cannot call setState while rendering.');
      }

      currentState[hookIndex] =
        valueOrFunction instanceof Function
          ? valueOrFunction(currentState[hookIndex])
          : valueOrFunction;
    }

    reconciler.currentHookIndex++;

    return [currentState[hookIndex], stateSetter];
  }
  return useState;
}
