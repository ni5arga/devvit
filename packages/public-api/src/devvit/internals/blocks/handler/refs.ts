import { _activeRenderContext } from './BlocksHandler.js';
import type { HookRef } from './types.js';

/**
 * This function is mostly used for tests.  It's a way to capture the hook id without changing code too much
 *
 * Example:
 *
 * ```tsx
 * const ref: HookRef = {};
 * let [count, setCount] = captureHookRef(useState(0), ref);
 * ```
 *
 * @param value
 * @param ref
 * @returns
 */
export function captureHookRef<T>(value: T, ref: HookRef): T {
  if (!_activeRenderContext) {
    throw new Error('No active render context');
  }

  if (typeof value === 'function') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (value as any)['captureHookRef'] = () => {
      if (!_activeRenderContext) {
        throw new Error('No active render context');
      }
      const id = _activeRenderContext._prevHookId;
      ref.id = id;
    };
  } else {
    const id = _activeRenderContext._prevHookId;
    if (!id) {
      throw new Error('No hook id found');
    }
    ref.id = id;
  }

  return value;
}
