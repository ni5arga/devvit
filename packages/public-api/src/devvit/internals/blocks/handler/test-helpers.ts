import type { UIEvent, UIRequest } from '@devvit/protos';
import { Header } from '@devvit/shared-types/Header.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import { _latestBlocksHandler } from './BlocksHandler.js';
import type { HookRef } from './types.js';

export const findHookId = (ref: HookRef): string => {
  return ref.id!;
};

export const findHookState = (ref: HookRef): JSONValue => {
  return _latestBlocksHandler!._latestRenderContext!._state[ref.id!];
};

export const EmptyRequest: UIRequest = { events: [] };

export const generatePressRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    userAction: {
      actionId: ref.id!,
    },
  };
  return {
    state: _latestBlocksHandler?._latestRenderContext?._state ?? {},
    events: [event],
  };
};

export const generateTimerRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    timer: {},
  };
  return {
    state: _latestBlocksHandler?._latestRenderContext?._state ?? {},
    events: [event],
  };
};

export const mockMetadata = {
  [Header.AppUser]: {
    values: ['t2_appuser'],
  },
  [Header.Subreddit]: {
    values: ['t5_devvit'],
  },
  [Header.User]: {
    values: ['t2_user'],
  },
};
