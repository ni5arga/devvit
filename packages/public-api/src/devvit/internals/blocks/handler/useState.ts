import { EffectType } from '@devvit/protos';
import type { JSONValue } from '@devvit/shared-types/json.js';
import type {
  AsyncUseStateInitializer,
  SetStateAction,
  UseStateInitializer,
  UseStateResult,
} from '../../../../types/hooks.js';
import { registerHook } from './BlocksHandler.js';
import type { RenderContext } from './RenderContext.js';
import type { Hook, HookParams } from './types.js';
import { RenderInterruptError } from './types.js';

export function useState(initialState: UseStateInitializer<boolean>): UseStateResult<boolean>;
export function useState(initialState: UseStateInitializer<number>): UseStateResult<number>;
export function useState(initialState: UseStateInitializer<string>): UseStateResult<string>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useState' }, ({ changed }) => {
    const state = initialState instanceof Function ? initialState() : initialState;
    if (state instanceof Promise) {
      throw new Error('Cannot use async initializer with useState, use useAsyncState instead.');
    }
    const setter = (action: SetStateAction<S>): void => {
      hook.state = action instanceof Function ? action(hook.state) : action;
      changed();
    };
    return { state, setter };
  });
  return [hook.state, hook.setter];
}

class AsyncStateHook<S extends JSONValue> implements Hook {
  state: { value: S | null; loading: boolean } = { value: null, loading: false };
  #changed: () => void;
  #initializer: AsyncUseStateInitializer<S>;
  #hookId: string;

  constructor(initializer: AsyncUseStateInitializer<S>, params: HookParams) {
    this.#initializer = initializer;
    this.#hookId = params.hookId;
    this.#changed = params.changed;
  }

  async onUIEvent(): Promise<void> {
    if (this.state.value === null) {
      this.state.value = await this.#initializer();
      this.state.loading = false;
      this.#changed();
    }
  }

  setter(action: SetStateAction<S>): void {
    this.state.value = action instanceof Function ? action(this.state.value as S) : action;
    this.#changed();
  }

  onLoad(renderContext: RenderContext): void {
    if (this.state.value === null && !this.state.loading) {
      this.state.loading = true;
      this.#changed();
      renderContext.emitEffect(this.#hookId, {
        type: EffectType.EFFECT_SEND_EVENT,
        // This is on the main queue, because it blocks app load.
        sendEvent: { event: { asyncRequest: { requestId: this.#hookId }, hook: this.#hookId } },
      });
      throw new RenderInterruptError();
    }
  }
}

export function useAsyncState<S extends JSONValue>(
  initializer: AsyncUseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useAsyncState' }, (params) => {
    return new AsyncStateHook(initializer, params);
  });

  return [hook.state.value as S, hook.setter.bind(hook)];
}
