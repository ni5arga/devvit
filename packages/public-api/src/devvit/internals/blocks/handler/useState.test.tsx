/** @jsx Devvit.createElement */
/** @jsxFrag Devvit.Fragment */

import type { UIRequest } from '@devvit/protos';
import { describe, expect, test } from 'vitest';
import type { UseStateResult } from '../../../../index.js';
import { Devvit } from '../../../Devvit.js';
import { BlocksHandler } from './BlocksHandler.js';
import { captureHookRef } from './refs.js';
import { EmptyRequest, findHookState, generatePressRequest, mockMetadata } from './test-helpers.js';
import type { HookRef } from './types.js';
import { useAsyncState, useState } from './useState.js';

const syncLegacyComponent: Devvit.BlockComponent = (
  _props: JSX.Props,
  { useState }: Devvit.Context
) => {
  const [state] = useState({ foo: 'bar' });
  return <text>{state.foo}</text>;
};

const deprecatedComponent: Devvit.BlockComponent = (
  _props: JSX.Props,
  { useState }: Devvit.Context
) => {
  const [state] = useState(async () => ({ foo: 'bar' }));
  return <text>{state.foo}</text>;
};

const conditionalComponent: Devvit.BlockComponent = () => {
  const [counter, setCounter] = captureHookRef(useState(0), counterRef);

  let fooValue = '';

  if (counter === 0) {
    const [foo] = useState('bar');
    fooValue = foo;
  }

  return (
    <vstack>
      <text>{counter}</text>
      <text>{fooValue}</text>
      <button
        onPress={() => {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          setCounter((counter) => (counter > 10 ? (undefined as any) : counter + 1));
        }}
      >
        Increment
      </button>
    </vstack>
  );
};

const multiComponent: Devvit.BlockComponent = () => {
  const [counter] = useState(0);
  const [foo] = useState('bar');
  const [baz] = useState({
    hello: 'world',
  });

  return (
    <vstack>
      <text>{counter}</text>
      <text>{foo}</text>
      <text>{baz.hello}</text>
    </vstack>
  );
};

const asyncComponent: Devvit.BlockComponent = () => {
  const [state] = useAsyncState(async () => {
    return 'hello world';
  });
  return <text>{state}!!</text>;
};

const syncImportComponent: Devvit.BlockComponent = () => {
  const [state] = useState({ foo: 'bar' });
  return <text>{state.foo}</text>;
};

const counterRef: HookRef = {};
const counterButtonRef: HookRef = {};

const counterComponent: Devvit.BlockComponent = () => {
  const [counter, setCounter] = captureHookRef(useState(0), counterRef);
  return (
    <vstack>
      <text>{counter}</text>
      <button
        onPress={captureHookRef(() => setCounter((counter) => counter + 1), counterButtonRef)}
      >
        Increment
      </button>
    </vstack>
  );
};

describe('useState', () => {
  describe('initial state', () => {
    test.each([syncImportComponent, syncLegacyComponent])(
      'uses the given initial state',
      async (c) => {
        const handler = new BlocksHandler(c);
        const response = await handler.handle(EmptyRequest, mockMetadata);

        expect(response.state).toMatchSnapshot();
        expect(JSON.stringify(response.state)).toContain('foo');
        expect(JSON.stringify(response.state)).toContain('bar');
      }
    );

    test('async deprecation displays message', async () => {
      const handler = new BlocksHandler(deprecatedComponent);
      let yay = false;
      try {
        await handler.handle(EmptyRequest, mockMetadata);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (e: any) {
        expect(e.message).toContain('useAsyncState instead');
        yay = true;
      }
      expect(yay).toBeTruthy();
    });

    test('async initializer resolves', async () => {
      const handler = new BlocksHandler(asyncComponent);
      const response = await handler.handle(EmptyRequest, mockMetadata);

      expect(response.state).toMatchSnapshot();
      expect(response.events.length).toEqual(1);
      expect(response.blocks).toBeUndefined();

      const event = response.events[0];
      expect(event?.async).toBeFalsy();
      const next: UIRequest = {
        events: [event],
        state: response.state,
      };
      const nextResponse = await handler.handle(next, mockMetadata);
      expect(nextResponse.events.length).toEqual(0);
      expect(JSON.stringify(nextResponse.blocks ?? '')).toContain('hello world');
    });
  });
});

describe('state setter', () => {
  test('updates the value given a new value', async () => {
    const handler = new BlocksHandler(counterComponent);
    await handler.handle(EmptyRequest, mockMetadata);
    expect(findHookState(counterRef)).toEqual(0);
    let req = generatePressRequest(counterButtonRef);
    await handler.handle(req, mockMetadata);
    expect(findHookState(counterRef)).toEqual(1);
    req = generatePressRequest(counterButtonRef);
    await handler.handle(req, mockMetadata);
    expect(findHookState(counterRef)).toEqual(2);
    req = generatePressRequest(counterButtonRef);
    await handler.handle(req, mockMetadata);
    expect(findHookState(counterRef)).toEqual(3);
  });

  test('can be used multiple times in a component', async () => {
    const handler = new BlocksHandler(multiComponent);
    const response = await handler.handle(EmptyRequest, mockMetadata);
    expect(response.state).toMatchSnapshot();
    expect(JSON.stringify(response.state)).toContain('0');
    expect(JSON.stringify(response.state)).toContain('bar');
    expect(JSON.stringify(response.state)).toContain('world');
  });

  test('must not be used outside of a component', async () => {
    expect(() => useState(0)).toThrowError();
  });

  test('must not be used conditionally', async () => {
    // No longer enforced in code.
  });
  test('handles void or undefined values properly', async () => {
    const handler = new BlocksHandler(conditionalComponent);
    await handler.handle(EmptyRequest, mockMetadata);
    expect(findHookState(counterRef)).toEqual(0);
    for (let i = 0; i < 15; i++) {
      const req = generatePressRequest(counterButtonRef);
      await handler.handle(req, mockMetadata);
    }
    expect(findHookState(counterRef)).toEqual(0);
  });

  test('typing is intuitive', () => {
    const _foo: Devvit.BlockComponent = (_props: JSX.Props) => {
      type Foo = {
        bar: string;
      };
      const foo: Foo = { bar: 'abc' };

      useState([1]) satisfies UseStateResult<number[]>;

      // @ts-expect-error
      useState(undefined);
      useState('abc');
      useState(1);
      useState(false);
      useState(null);
      useState([]);
      useState({});
      useState({ a: 123 }) satisfies UseStateResult<{ a: number }>;
      useState([{ a: 123 }, 1, null]);
      useState(foo);
      // @ts-expect-error
      useState(() => undefined);
      useState(() => 'abc');
      useState(() => 1);
      useState(() => false);
      useState(() => null);
      useState(() => []);
      useState(() => ({}));
      useState(() => foo);
      // @ts-expect-error
      useState(Promise.resolve(null));
      // @ts-expect-error
      useState(() => Promise.resolve(undefined));
      // @ts-expect-error
      useState(() => Promise.resolve('abc'));
      // @ts-expect-error
      useState(() => Promise.resolve(1));
      // @ts-expect-error
      useState(() => Promise.resolve(false));
      // @ts-expect-error
      useState(() => Promise.resolve(null));
      // @ts-expect-error
      useState(() => Promise.resolve([]));
      // @ts-expect-error
      useState(() => Promise.resolve({}));
      // @ts-expect-error
      useState(() => Promise.resolve(foo));
      // @ts-expect-error
      useState(new (class {})());
      // @ts-expect-error
      useState(new Map());
      // @ts-expect-error
      useState(new Set());
      // @ts-expect-error
      useState(new Date());
      // @ts-expect-error
      useState(() => Promise.resolve(new (class {})()));
      // @ts-expect-error
      useState(() => Promise.resolve(new Map()));
      // @ts-expect-error
      useState(() => Promise.resolve(new Set()));
      // @ts-expect-error
      useState(() => Promise.resolve(new Date()));
      return <text />;
    };
  });
});
