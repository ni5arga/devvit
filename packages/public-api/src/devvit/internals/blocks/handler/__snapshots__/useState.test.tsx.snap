// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`state setter > can be used multiple times in a component 1`] = `
{
  "multiComponent.useState-0": 0,
  "multiComponent.useState-1": "bar",
  "multiComponent.useState-2": {
    "hello": "world",
  },
}
`;

exports[`useState > initial state > async initializer resolves 1`] = `
{
  "asyncComponent.useAsyncState-0": {
    "loading": true,
    "value": null,
  },
}
`;

exports[`useState > initial state > uses the given initial state 1`] = `
{
  "syncImportComponent.useState-0": {
    "foo": "bar",
  },
}
`;

exports[`useState > initial state > uses the given initial state 2`] = `
{
  "syncLegacyComponent.useState-0": {
    "foo": "bar",
  },
}
`;
